src/core/coap/coap.cpp

```diff
@@ -1011,10 +1011,10 @@ Message *CoapBase::FindRelatedRequest(const Message          &aResponse,
     {
         aMetadata.ReadFrom(message);
 
-        if (((aMetadata.mDestinationAddress == aMessageInfo.GetPeerAddr()) ||
+        if (((aMetadata.mDestinationAddress == aMessageInfo.GetSockAddr()) ||
              aMetadata.mDestinationAddress.IsMulticast() ||
              aMetadata.mDestinationAddress.GetIid().IsAnycastLocator()) &&
-            (aMetadata.mDestinationPort == aMessageInfo.GetPeerPort()))
+            (aMetadata.mDestinationPort == aMessageInfo.GetSockPort()))
```

Comment:
I think the existing code is correct. It leverages `MessageInfo`, which is applicable to both transmitting (tx) and receiving (rx) messages, dynamically adjusting the interpretation of "peer" and "local".

Within a `MessageInfo` the peer address/port consistently refers to the other device involved in the communication (either the source we received the message from or the destination we are sending it to).

Specifically:
- For a received message, `GetPeerAddress()` will be the source address (sender's address).
- For a sent message, `GetPeerAddress()` will be the destination address.